File: Embarcadero/Berlin/CPP/NotesResource/NotesAdapterModuleU.cpp
//---------------------------------------------------------------------------

// This software is Copyright (c) 2015 Embarcadero Technologies, Inc.
// You may only use this software if you are an authorized licensee
// of an Embarcadero developer tools product.
// This software is considered a Redistributable as defined under
// the software license agreement that comes with the Embarcadero Products
// and is subject to that software license agreement.

//---------------------------------------------------------------------------

#pragma hdrstop

#include "NotesAdapterModuleU.h"
#include "NotesClientModuleU.h"
#include <FMX.Dialogs.hpp>
#include <REST.Types.hpp>
#include <vector>
#include <memory>
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma classgroup "FMX.Controls.TControl"
#pragma resource "*.dfm"

TNotesAdapterModule *NotesAdapterModule;
//---------------------------------------------------------------------------
__fastcall TNotesAdapterModule::TNotesAdapterModule(TComponent* Owner)
	: TDataModule(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TNotesAdapterModule::AfterPost(TBindSourceAdapter * Sender)
{
	TNoteWrapper * lWrapper = (TNoteWrapper*)FBindSourceAdapter->List->Items[FBindSourceAdapter->ItemIndex];
	try
	{
		if(lWrapper->ID.IsEmpty()) {
			String lID = "";
			NotesClientModule->AddNote(lWrapper->GetNote(), lID);
			lWrapper->SetID(lID);
		} else {
			NotesClientModule->UpdateNote(lWrapper->GetNote());
		}
	} catch(...) {
		FBindSourceAdapter->Edit();
		throw;
	}
}
//---------------------------------------------------------------------------
void __fastcall TNotesAdapterModule::BeforeDelete(TBindSourceAdapter * Sender)
{
	TNoteWrapper * lWrapper = (TNoteWrapper*)FBindSourceAdapter->List->Items[FBindSourceAdapter->ItemIndex];
	if(!lWrapper->ID.IsEmpty()) {
        	NotesClientModule->DeleteNote(lWrapper->ID);
    }
}
//---------------------------------------------------------------------------
TBindSourceAdapter* __fastcall TNotesAdapterModule::GetBindSourceAdapter(void)
{
	if(FBindSourceAdapter == NULL) {
		TList__1<TObject*> * LList = new TList__1<TObject*>();
		FBindSourceAdapter = new TListBindSourceAdapter(this, LList, __classid(TNoteWrapper), true);
		FBindSourceAdapter->AfterPost = &AfterPost;
		FBindSourceAdapter->BeforeDelete = &BeforeDelete;
	}
	return FBindSourceAdapter;
}
//---------------------------------------------------------------------------
void __fastcall TNotesAdapterModule::UpdateAdapter(std::vector<TNote*> *ANotes)
{
	if (ComponentState.Contains(csDestroying)) {
		return;
	}
	TList__1<TObject*> *lList = new TList__1<TObject*>();
	try {
		for(std::vector<TNote*>::iterator it = ANotes->begin(); it != ANotes->end(); it++)
		{
			lList->Add(new TNoteWrapper(*it));
                }
                if(FBindSourceAdapter == NULL) {
                        FBindSourceAdapter = new TListBindSourceAdapter(this, lList, __classid(TNoteWrapper), true);
                }
                else {
                	FBindSourceAdapter->SetList(lList, true);
                }
                FBindSourceAdapter->Active = true;
	} catch (...) {
		FreeAndNil(lList);
		throw;
	}
}
//---------------------------------------------------------------------------
void __fastcall TNotesAdapterModule::RefreshAdapter(void)
{
	std::vector<TNote*> lNotes = NotesClientModule->GetNotes();
	UpdateAdapter(&lNotes);
}
//---------------------------------------------------------------------------

File: Embarcadero/Rio/Object Pascal/NotesResource/NotesAdapterModuleU.pas
//---------------------------------------------------------------------------

// This software is Copyright (c) 2015 Embarcadero Technologies, Inc.
// You may only use this software if you are an authorized licensee
// of an Embarcadero developer tools product.
// This software is considered a Redistributable as defined under
// the software license agreement that comes with the Embarcadero Products
// and is subject to that software license agreement.

//---------------------------------------------------------------------------

unit NotesAdapterModuleU;

interface

uses
  System.SysUtils, System.Generics.Collections, System.Classes, Data.Bind.ObjectScope,
  NoteTypesU;

type
  // Expose data binding properties
  TNoteWrapper = class
  private
    FNote: TNote;
    function GetContent: string;
    function GetID: string;
    function GetTitle: string;
    procedure SetContent(const Value: string);
    procedure SetTitle(const Value: string);
  public
    constructor Create(const ANote: TNote);
    function GetNote: TNote;
    procedure SetID(const AID: string);
    property Title: string read GetTitle write SetTitle;
    property Content: string read GetContent write SetContent;
    property ID: string read GetID;
  end;

  TNotesAdapterModule = class(TDataModule)
  private
    FBindSourceAdapter: TListBindSourceAdapter<TNoteWrapper>;
    procedure AfterPost(Sender: TBindSourceAdapter);
    procedure BeforeDelete(Sender: TBindSourceAdapter);
  public
    function GetBindSourceAdapter: TBindSourceAdapter;
    procedure UpdateAdapter(const ANotes: TArray<TNote>);
    { Private declarations }
  public
    { Public declarations }
    procedure RefreshAdapter; overload;
    property BindSourceAdapter: TBindSourceAdapter read GetBindSourceAdapter;
  end;

var
  NotesAdapterModule: TNotesAdapterModule;

implementation

{%CLASSGROUP 'FMX.Controls.TControl'}

uses NotesClientModuleU;

{$R *.dfm}

procedure TNotesAdapterModule.RefreshAdapter;
var
  LNotes: TArray<TNote>;
begin
  LNotes := NotesClientModule.GetNotes;
  UpdateAdapter(LNotes);
end;

procedure TNotesAdapterModule.UpdateAdapter(const ANotes: TArray<TNote>);
var
  LList: TList<TNoteWrapper>;
  LNote: TNote;
begin
  if csDestroying in ComponentState then
    Exit;
  LList := TObjectList<TNoteWrapper>.Create;
  try
    for LNote in ANotes do
      LList.Add(TNoteWrapper.Create(LNote));
    if FBindSourceAdapter = nil then
      FBindSourceAdapter := TListBindSourceAdapter<TNoteWrapper>.Create(Self, LList, True)
    else
      FBindSourceAdapter.SetList(LList, True);
    FBindSourceAdapter.Active := True;
  except
    LList.Free;
    raise;
  end;
end;

function TNotesAdapterModule.GetBindSourceAdapter: TBindSourceAdapter;
begin
  if FBindSourceAdapter = nil then
  begin
     // Create empty adapter
     FBindSourceAdapter := TListBindSourceAdapter<TNoteWrapper>.Create(Self, TList<TNoteWrapper>.Create, True);
     FBindSourceAdapter.AfterPost := AfterPost;
     FBindSourceAdapter.BeforeDelete := BeforeDelete;
  end;
  Result := FBindSourceAdapter;
end;

procedure TNotesAdapterModule.BeforeDelete(Sender: TBindSourceAdapter);
var
  LWrapper: TNoteWrapper;
begin
  LWrapper := FBindSourceAdapter.List[FBindSourceAdapter.ItemIndex];
  if LWrapper.ID <> '' then
    NotesClientModule.DeleteNote(LWrapper.ID);
end;

procedure TNotesAdapterModule.AfterPost(Sender: TBindSourceAdapter);
var
  LWrapper: TNoteWrapper;
  LID: string;
begin
  try
    LWrapper := FBindSourceAdapter.List[FBindSourceAdapter.ItemIndex];
    if LWrapper.ID = '' then
    begin
      NotesClientModule.AddNote(LWrapper.GetNote, LID);
      LWrapper.SetID(LID);
    end
    else
      NotesClientModule.UpdateNote(LWrapper.GetNote);
  except
    FBindSourceAdapter.Edit;  // Return to edit mode
    raise;
  end;
end;

{ TRecipe }

constructor TNoteWrapper.Create(const ANote: TNote);
begin
  FNote := ANote;
end;

function TNoteWrapper.GetContent: string;
begin
  Result := FNote.Content;
end;

function TNoteWrapper.GetID: string;
begin
  Result := FNote.ID;
end;

function TNoteWrapper.GetNote: TNote;
begin
  Result := FNote;
end;

function TNoteWrapper.GetTitle: string;
begin
  Result := FNote.Title;

end;

procedure TNoteWrapper.SetContent(const Value: string);
begin
  FNote.Content := Value;
end;

procedure TNoteWrapper.SetID(const AID: string);
begin
  FNote.ID := AID;
end;

procedure TNoteWrapper.SetTitle(const Value: string);
begin
  FNote.Title := Value;
end;

end.
